collections/iter.go

```diff
@@ -0,0 +1,298 @@
+	r.end = BoundInclusive(end)
+	return r
+}
+
+// EndExclusive makes the range contain only keys which are smaller to the provided end K.
+func (r *Range[K]) EndExclusive(end K) *Range[K] {
+	r.end = BoundExclusive(end)
+	return r
+}
+
+func (r *Range[K]) Descending() *Range[K] {
+	r.order = OrderDescending
+	return r
+}
+
+func (r *Range[K]) RangeValues() (prefix *K, start *Bound[K], end *Bound[K], order Order, err error) {
+	return r.prefix, r.start, r.end, r.order, nil
+}
+
+// iteratorFromRanger generates an Iterator instance, with the proper prefixing and ranging.
+// a nil Ranger can be seen as an ascending iteration over all the possible keys.
+func iteratorFromRanger[K, V any](ctx context.Context, m Map[K, V], r Ranger[K]) (iter Iterator[K, V], err error) {
+	// defaults
+	var (
+		prefix *K
+		start  *Bound[K]
+		end    *Bound[K]
+		order  = OrderAscending
+	)
+	// override defaults only if a Ranger is provided.
+	if r != nil {
+		prefix, start, end, order, err = r.RangeValues()
+		if err != nil {
+			return iter, err
+		}
+	}
+	var prefixBytes []byte
+	if prefix != nil {
+		prefixBytes, err = encodeKeyWithPrefix(m.prefix, m.kc, *prefix)
+		if err != nil {
+			return iter, err
+		}
+	} else {
+		prefixBytes = m.prefix
+	}
+	var startBytes []byte // default is nil
+	if start != nil {
+		startBytes, err = encodeKeyWithPrefix(nil, m.kc, start.value)
+		if err != nil {
+			return
+		}
+		// iterators are inclusive at start by default
+		// so if we want to make the iteration exclusive
+		// we extend by one byte.
+		if !start.inclusive {
+			startBytes = extendOneByte(startBytes)
+		}
+	}
+	var endBytes []byte // default is nil
+	if end != nil {
+		endBytes, err = encodeKeyWithPrefix(nil, m.kc, end.value)
+		if err != nil {
+			return iter, err
+		}
+		// iterators are exclusive at end by default
+		// so if we want to make the iteration
+		// inclusive we need to extend by one byte.
+		if end.inclusive {
+			endBytes = extendOneByte(endBytes)
+		}
+	}
+
+	store, err := m.getStore(ctx)
+	if err != nil {
+		return iter, err
+	}
+	prefixedStartBytes := append(prefixBytes, startBytes...)
+	// since we're dealing with a prefixed end, if end bytes
+	// are not specified then we need to increase the prefix
+	// by 1, otherwise we would end up ranging over [bytes, bytes),
+	// which is invalid.
+	var prefixedEndBytes []byte
+	if endBytes == nil {
+		prefixedEndBytes = storetypes.PrefixEndBytes(prefixBytes)
+	} else {
+		prefixedEndBytes = append(prefixBytes, endBytes...)
+	}
+
+	var storeIter storetypes.Iterator
+	switch order {
+	case OrderAscending:
+		storeIter = store.Iterator(prefixedStartBytes, prefixedEndBytes)
+	case OrderDescending:
+		storeIter = store.ReverseIterator(prefixedStartBytes, prefixedEndBytes)
+	default:
+		return iter, fmt.Errorf("collections: unrecognized order identifier: %d", order)
+	}
+
+	iter.kc = m.kc
+	iter.vc = m.vc
+	iter.iter = storeIter
+	iter.prefixLength = len(m.prefix)
+
+	return iter, nil
```

Comment:
IDK if here we should add safety guarantees, eg if !iter.Valid() then error, some implementations of iterator just panic.

Reply:
Modifies to return an `ErrInvalidIterator` in case the iteration yields to an invalid iterator.